### ISSUE TITLE
static_path causes major slowdown 

### ISSUE THREAD

**Issue by ndeet** [phoenixframework/phoenix#1451]
On my [testapp](https://github.com/ndeet/pete) it seems that the ```static_path``` is the issue causing the server load not to grow >6 on a 12 core server and only giving 3000 req/s throughput.

I used this in my template
```
<img src="<%= static_path(@conn, "/images/") <> elem(@image, 0) <> ".jpg" %>" />
```

@chrismccord you removed that in your PR and changed it to a absolute path:
```
<img src="/images/<%= elem(@image, 0) <> ".jpg" %>" />
```

12-core vm using static link:
```
root@ubuntu-8gb-fra1-01:~/wrk-4.0.1# ./wrk -t12 -c100 -d60s --timeout 1000 http://46.101.247.95:4001/gallery
Running 1m test @ http://46.101.247.95:4001/gallery
  12 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    21.88ms    5.13ms  80.43ms   74.67%
    Req/Sec   367.33     51.74   525.00     71.15%
  263609 requests in 1.00m, 493.29MB read
Requests/sec:   4387.46
Transfer/sec:      8.21MB
```

without (I had tears in my eyes):
```
root@ubuntu-8gb-fra1-01:~/wrk-4.0.1# ./wrk -t12 -c100 -d60s --timeout 1000 http://46.101.247.95:4001/gallery
Running 1m test @ http://46.101.247.95:4001/gallery
  12 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.14ms    2.48ms  79.44ms   90.97%
    Req/Sec     2.71k   512.88     4.70k    68.53%
  1943797 requests in 1.00m, 3.57GB read
Requests/sec:  32345.54
Transfer/sec:     60.77MB
```

Side note: as talked in the irc I now use Erlang 18.1, which seems to push the former measured 3000 req/s to 4387 req/s - but can't say for sure as the VMs may differ in performance. But nevertheless the big bottleneck seems to be the `static_path`

**Comment by ndeet** [comment]
I just cross checked to be sure it is not an routing or DNS or other weird issue and replaced the absolute path with URL to IP and a local domain in the image tag, and the results are the same, so it seems very likely to be the ```static_path``` function.

**Comment by chrismccord** [comment]
> @chrismccord you removed that in your PR and changed it to a absolute path:

This should have been fast, but I noticed the other framework had a hard-coded path so I removed the call. We do an ets cache lookup, which be super cheap, but I will investigate tomorrow on possible bottlenecks.

Also as a sanity check, can you verify that you isolated this huge difference to `static_path`? The PR I sent also stripped out other plugs, so I want to double check if the gains could be elsewhere.

**Comment by ndeet** [comment]
I'm pretty confident that it is caused by ```static_path```, you can easily verify on your side. After your pull request on my localhost:
```
$ wrk -t4 -c100 -d60s --timeout 1000 http://localhost:4001/gallery
Running 1m test @ http://localhost:4001/gallery
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     8.00ms   12.69ms 262.59ms   94.68%
    Req/Sec     3.50k     1.30k    7.95k    66.98%
  831665 requests in 1.00m, 1.53GB read
Requests/sec:  13842.43
Transfer/sec:     26.01MB
```

If I add _nothing_ back from your PR but only the static_path function in the template:
```
$ wrk -t4 -c100 -d60s --timeout 1000 http://localhost:4001/gallery
Running 1m test @ http://localhost:4001/gallery
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    28.31ms    4.75ms 144.52ms   95.16%
    Req/Sec     0.89k    89.94     1.01k    81.22%
  212681 requests in 1.00m, 399.61MB read
Requests/sec:   3540.97
Transfer/sec:      6.65MB
```

I can do the reverse check and add everything back except the static_path, but even if other plugs somehow slow things down the above test is evidence enough that it is static_path?

**Comment by josevalim** [comment]
We are doing file system lookup when there are no entries in the ETS table,
which will always be the case for "/images/". You should always lookup the
full thing. Chris, tomorrow we should talk about the proper fix. :)

On Wednesday, January 6, 2016, Andreas Tasch <notifications@github.com>
wrote:

> I'm pretty confident that it is caused by static_path, you can easily
> verify on your side. After your pull request on my localhost:
>
> $ wrk -t4 -c100 -d60s --timeout 1000 http://localhost:4001/gallery
> Running 1m test @ http://localhost:4001/gallery
>   4 threads and 100 connections
>   Thread Stats   Avg      Stdev     Max   +/- Stdev
>     Latency     8.00ms   12.69ms 262.59ms   94.68%
>     Req/Sec     3.50k     1.30k    7.95k    66.98%
>   831665 requests in 1.00m, 1.53GB read
> Requests/sec:  13842.43
> Transfer/sec:     26.01MB
>
> If I add *nothing* back from your PR but only the static_path function in
> the template:
>
> $ wrk -t4 -c100 -d60s --timeout 1000 http://localhost:4001/gallery
> Running 1m test @ http://localhost:4001/gallery
>   4 threads and 100 connections
>   Thread Stats   Avg      Stdev     Max   +/- Stdev
>     Latency    28.31ms    4.75ms 144.52ms   95.16%
>     Req/Sec     0.89k    89.94     1.01k    81.22%
>   212681 requests in 1.00m, 399.61MB read
> Requests/sec:   3540.97
> Transfer/sec:      6.65MB
>
> I can do the reverse check and add everything back except the static_path,
> but even if other plugs somehow slow things down the above test is evidence
> enough that it is static_path?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phoenixframework/phoenix/issues/1451#issuecomment-169275878>
> .
>


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by ndeet** [comment]
Ok did the reverse check, so I added everything back you removed:
```
$ wrk -t4 -c100 -d60s --timeout 1000 http://localhost:4001/gallery
Running 1m test @ http://localhost:4001/gallery
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    31.21ms    7.60ms 185.48ms   93.87%
    Req/Sec   811.47    108.87     1.00k    83.10%
  193768 requests in 1.00m, 372.94MB read
Requests/sec:   3224.88
Transfer/sec:      6.21MB
```

Now, removing only ```static_path``` in the image.html.eex:
```
$ wrk -t4 -c100 -d60s --timeout 1000 http://localhost:4001/gallery
Running 1m test @ http://localhost:4001/gallery
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     8.79ms   10.19ms 222.28ms   94.77%
    Req/Sec     3.13k   416.36     4.92k    75.79%
  745257 requests in 1.00m, 1.40GB read
Requests/sec:  12401.09
Transfer/sec:     23.87MB
```

So, yes the other plugs slow things down ~1400 req/s but the ```static_path``` seems to take away ~9100 req/s.

**Comment by josevalim** [comment]
Yup, it is not that `static_path` is slowing things down, `static_path` is straight a bottleneck. It doesn't matter how fast or slow you make other things, you ain't gonna pass 3000req/s. :) Great find @ndeet!

**Comment by ndeet** [comment]
Thank you both for helping me to isolate the issue.

Out of curiosity I did another experiment to verify:
```
mix phoenix.new hello
# compile and run prod server
```

run wrk:
$ wrk -t4 -c100 -d30s --timeout 1000 http://localhost:4001
Running 30s test @ http://localhost:4001
  4 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    18.71ms    3.65ms  82.34ms   89.67%
    Req/Sec     1.35k   109.03     1.93k    89.25%
  160780 requests in 30.03s, 348.40MB read
Requests/sec:   5353.48
Transfer/sec:     11.60MB
```

Add 1 call to static_path to index.html.eex
```
<%= static_path(@conn, "/images/") %>
```

requests drop to ```Requests/sec:   3801.34````

add 4 calls to static_path requests drop to ```Requests/sec:   1878.65```

I guess it is best practice and the intention to use static_path on all assets in template files. So it is likely that there are many more calls to static_path in a standard webapp which will get ugly if you want to scale.

Thanks for fixing this, I am glad that I now can update the performance stats to show Phoenix full potential and as I expected it is way faster than interpreted PHP even if the framework is compiled in C.

@josevalim Yesterday I also recompiled back to Erlang 18.2.1 and guess what the in the first post mentioned **32_345.54** req/s increased to **39_000** req/s. So 18.2.1 improved >20%.
I then wanted to literally break phoenix by letting wrk doing 48 threads with 500 connections. No error and the req/s went up to 42_000 (I guess now I hit the bandwidth limit with ~80mb/s). This is totally insane performance :) The load was still at 8 but CPUs got much more work done.

**Comment by chrismccord** [comment]
@ndeet What José is saying is the slowdown is caused by your calls to static_path which do not provide the full file path, i.e. `static_path(conn, "/images/") <> file`. This causes us to hit disk doing a lookup on /images/ on every request, where the standard usage `static_path(conn, "/images/#{file}")` would have been cached. We definitely want to take care of the former case, but to be clear static_path usage with a relative file path that exists will be super fast. Thanks for the detailed info!

**Comment by ndeet** [comment]
Ahh, ok I guess I understand the issue now. The problem was that (due to my lack of understanding the correct syntax) I used static_path to get the path only (which seems to be slow) and after that I concatenated the string to the full path.:
```
<img src="<%= static_path(@conn, "/images/") <> elem(@image, 0) <> ".jpg" %>" />
``` 

The correct and way faster syntax is this(?): (successfully tested this with wrk)
```
<img src="<%= static_path(@conn, "/images/#{elem(@image, 0)}.jpg") %>" />
```

So if it is not possible to make the path only lookup as fast it may make sense to note it somewhere in the docs? And maybe an example how to correctly use static_path with static assets and variables? Let me know where it fits and I do a PR for the docs for that.
